c_src/membrane_audio_mix_plugin/mixer.c

```diff
@@ -1,210 +1,246 @@
+                uint32_t *samples_length, UnifexState *state) {
+  if (values_length == 0 || !values) {
+    return;
+  }
+
+  bool is_wave_positive = state->is_wave_positive;
+  uint32_t start = 0;
+  uint32_t end = next_sign_change(values, values_length, is_wave_positive);
+
+  uint32_t samples_size = 0;
+  while (end < values_length) {
+    uint32_t queue_length = state->queue_length;
+    get_samples(samples + samples_size, values + start, end - start, state);
+    samples_size += (end - start + queue_length) * state->sample_size;
+    start = end;
+    is_wave_positive = !is_wave_positive;
+    end =
+        next_sign_change(values + end, values_length - end, is_wave_positive) +
+        end;
+  }
+
+  state->is_wave_positive = is_wave_positive;
+
+  uint32_t new_queue_length = state->queue_length + end - start;
+  int64_t *new_queue = unifex_alloc(new_queue_length * sizeof(int64_t));
+  memcpy(new_queue, state->queue, state->queue_length * sizeof(int64_t));
+  memcpy(new_queue + state->queue_length, values + start,
+         (end - start) * sizeof(int64_t));
+  state->queue_length = new_queue_length;
+  unifex_free(state->queue);
+  state->queue = new_queue;
+
+  *samples_length = samples_size / state->sample_size;
+}
 
-    if (is_last_wave)
-    {
-        get_samples(samples + samples_start, values + start, end - start, state);
-        start = end;
-    }
-    else
-    {
-        unsigned int new_queue_length = state->queue_length + end - start;
-        int64_t *new_queue = unifex_alloc(new_queue_length * sizeof(int64_t));
-        memcpy(new_queue, state->queue, state->queue_length * sizeof(int64_t));
-        memcpy(new_queue + state->queue_length, values + start, (end - start) * sizeof(int64_t));
-        state->queue_length = new_queue_length;
-        unifex_free(state->queue);
-        state->queue = new_queue;
+/**
+ *  Mixes `buffers` to one buffer.
+ */
+UNIFEX_TERM mix(UnifexEnv *env, UnifexPayload **buffers,
+                uint32_t buffers_length, UnifexState *state) {
+  uint32_t values_length = 0;
+  int64_t *values = NULL;
+
+  if (buffers_length > 0) {
+    uint32_t min_size = buffers[0]->size;
+    for (uint32_t i = 1; i < buffers_length; ++i) {
+      if (buffers[i]->size < min_size) {
+        min_size = buffers[i]->size;
+      }
     }
+    values_length = min_size / state->sample_size;
+    values = unifex_alloc(values_length * sizeof(int64_t));
+    get_values(buffers, buffers_length, values, values_length, state);
+  }
+
+  uint32_t samples_length = values_length + state->queue_length;
+  uint8_t *samples = unifex_alloc(samples_length * state->sample_size);
+
+  cut_values(values, values_length, samples, &samples_length, state);
+
+  unifex_free(values);
+  UnifexPayload out_payload;
+  uint32_t output_size = samples_length * state->sample_size;
+  unifex_payload_alloc(env, UNIFEX_PAYLOAD_BINARY, output_size, &out_payload);
+  memcpy(out_payload.data, samples, output_size);
+  unifex_free(samples);
```

Comment:
Why do you not store the samples right away in the output binary? You can use `unifex_payload_realloc` if it isn't fully filled